provider/ansible/version_added.rb

```diff
@@ -0,0 +1,121 @@
+# Copyright 2019 Google Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+CURRENT_ANSIBLE_VERSION = '2.8'.freeze
+
+module Provider
+  module Ansible
+    # All logic involved with altering the version_added yaml file and reading
+    # values from it.
+    module VersionAdded
+      def build_version_added
+        product_name = @api.name.downcase
+        versions_file = "products/#{product_name}/ansible_version_added.yaml"
+        raise 'File not found' unless File.exist?(versions_file)
+
+        versions = if File.exist?(versions_file)
+                     YAML.safe_load(File.read(versions_file), [Symbol])
+                   else
+                     {}
+                   end
+
+        struct = {
+          facts: {},
+          regular: {}
+        }
+
+        # Build out paths for regular modules.
+        @api.objects.reject(&:exclude).each do |obj|
+          next if obj.not_in_version?(@api.version_obj_or_default('ga'))
+
+          resource = {
+            version_added: correct_version([:regular, obj.name], versions)
+          }
+
+          # Add properties.
+          obj.all_user_properties.each do |prop|
+            resource[prop.name.to_sym] = property_version(prop, [:regular, obj.name], versions)
+          end
+          struct[:regular][obj.name.to_sym] = resource
+
+          # Add facts modules from facts datasources.
+          struct[:facts][obj.name.to_sym] = {
+            version_added: correct_version([:facts, obj.name], versions)
+          }
+        end
+
+        # Write back to disk.
+        File.write("products/#{product_name}/ansible_version_added.yaml", struct.to_yaml)
+
+        struct
+      end
+
+      def version_added(object, type = :regular)
+        if object.is_a?(Api::Resource)
```

Comment:
This function is used for fetching the version_added.

Unlike the config files, I try to be smarter about fetching version_added fields. The goal was to have a no-op with how I do it now. 

* No version_added if the version_added would match the module.
* Nested fields don't need a version_added if the higher-level field has a version_added.

Reply:
Maybe move this PR comment into a code comment.